### ISSUE TITLE
Terraform crash while creating plan

### ISSUE THREAD

**Issue by jim-leary** [hashicorp/terraform#31510]
### Terraform Version

```
Terraform v1.2.5
on darwin_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v3.19.0
+ provider registry.terraform.io/gavinbunney/kubectl v1.14.0
+ provider registry.terraform.io/hashicorp/aws v4.23.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.2.0
+ provider registry.terraform.io/hashicorp/github v4.27.1
+ provider registry.terraform.io/hashicorp/helm v2.6.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.12.1
+ provider registry.terraform.io/hashicorp/random v3.3.2
+ provider registry.terraform.io/hashicorp/tls v3.4.0
+ provider registry.terraform.io/petoju/mysql v3.0.17
```

### Terraform Configuration Files

```
 terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.0"

    }
    mysql = {
      source = "petoju/mysql"
      version = "~> 3.0"
    }
    kubernetes = {
      source  = "hashicorp/kubernetes"
      version = "~> 2.11"
    }
    kubectl = {
      source = "gavinbunney/kubectl"
      version = "~> 1.14"
    }
    cloudflare = {
      source = "cloudflare/cloudflare"
      version = "~> 3.18"
    }
  }
}
...
```

**Comment by jim-leary** [comment]
[terraform_debug.log.gz](https://github.com/hashicorp/terraform/files/9184759/terraform_debug.log.gz)

**Comment by jim-leary** [comment]
This was working correctly all day, then around 16:00 EDST I started getting plan errors.

**Comment by jim-leary** [comment]
[terraform_debug.log.gz](https://github.com/hashicorp/terraform/files/9184787/terraform_debug.log.gz)

**Comment by jbardin** [comment]
Hi @jim-leary,

Thanks for reporting the issue. The stack trace from the crash is here:
```
github.com/zclconf/go-cty/cty.Type.AttributeTypes(...)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/object_type.go:187
github.com/zclconf/go-cty/cty/convert.unifyObjectTypesToMap({0xc00231b480, 0x2, 0x3870f00?}, 0x1)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/convert/unify.go:368 +0x579
github.com/zclconf/go-cty/cty/convert.unifyTupleTypesToList({0xc00231b480, 0x2, 0x2}, 0x1)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/convert/unify.go:487 +0x40a
github.com/zclconf/go-cty/cty/convert.unifyTupleTypes({0xc00231b480?, 0x2, 0x2}, 0x1, 0x0?)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/convert/unify.go:418 +0x6b9
github.com/zclconf/go-cty/cty/convert.unify({0xc00231b480?, 0x2, 0x2}, 0x1)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/convert/unify.go:85 +0x41e
github.com/zclconf/go-cty/cty/convert.UnifyUnsafe(...)
        /Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.10.0/cty/convert/public.go:82
github.com/hashicorp/hcl/v2/hclsyntax.(*ConditionalExpr).Value(0xc00029d9d0, 0xc001ee7080)
        /Users/distiller/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.12.0/hclsyntax/expression.go:610 +0x2ab
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0xc00055c7e0, {0x386faa8?, 0xc00029d9d0}, {{0x3870ec8?, 0x4c53230?}})
        /Users/distiller/project/project/internal/lang/eval.go:171 +0x148
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateExpr(0x386faa8?, {0x386faa8, 0xc00029d9d0}, {{0x3870ec8?, 0x4c53230?}}, {0x0?, 0x0?})
        /Users/distiller/project/project/internal/terraform/eval_context_builtin.go:282 +0xc5
github.com/hashicorp/terraform/internal/terraform.(*NodeLocal).Execute(0x0?, {0x387f0e0, 0xc00207e540}, 0x60?)
        /Users/distiller/project/project/internal/terraform/node_local.go:153 +0x576
```

Can you supply a minimal example that causes the same crash so we can reproduce the issue?

**Comment by jim-leary** [comment]
Example of expected results:

```
# test-crash local example
# Jim Leary
# Example of bad code with expected failure

locals {
  
  default_lifecycle_rules = {
      id      = "default"
      enabled = true
      transition = [
        {
          days          = 1825
          storage_class = "GLACIER"
        }
      ]
      expiration = {
        days = 2555
      }
      noncurrent_version_expiration = {
        newer_noncurrent_versions = 3
        days = 3
      }
    }

  lifecycle_rules = {}

  # This should fail since the true/false values are different objects - and it does so without a crash
  all_lifecycle_rules = length(local.lifecycle_rules) >  0 ? [local.default_lifecycle_rules, local.lifecycle_rules] : [local.default_lifecycle_rules]

}

resource "null_resource" "test" {

    provisioner "local-exec" {
    command = "echo main test"
  }
}
```

**Comment by jim-leary** [comment]
I am able to reproduce the error with some simple code - it's rather interesting:

```
 local-module for test of crash
# Jim Leary
# local-module/main.tf


variable "lifecycle_rules" {
  description = "additional list of lifecycle rules maps"
  type        = any
  default     = {}
}


locals {

  default_lifecycle_rules = {
      id      = "default"
      enabled = true
      transition = [
        {
          days          = 1825
          storage_class = "GLACIER"
        }
      ]
      expiration = {
        days = 2555
      }
      noncurrent_version_expiration = {
        newer_noncurrent_versions = 3
        days = 3
      }
    }


  # This code will induce the crash
  all_lifecycle_rules = length(var.lifecycle_rules) >  0 ? [local.default_lifecycle_rules, var.lifecycle_rules] : [local.default_lifecycle_rules]

}


resource "null_resource" "test" {

    provisioner "local-exec" {
    command = "echo module local-module"
  }
}
```

When my root process calls the module above, it is not passing any variables and allowing the default values to be used.  The variable "lifecycle_rules" is set to type **any**, and the default value is **{}**.  This will cause the crash.  Changing the variable type to **map** eliminates the crash.  The key here is the use of a **variable**.  Using locals will not cause a crash.  Hope this helps.

**Comment by jim-leary** [comment]
Thanks - the original "any" designation was an bug on my part; however, not having Terraform panic and cause state-locks would be a big help.

**Comment by apparentlymart** [comment]
The upstream library change that @jbardin mentioned was included in one of the v1.2 patch releases and now in the v1.3 series, so this should now be a normal error explaining the problem instead of a panic. Therefore I'm going to close this.

If you see something which seems similar to this, please open a new issue for it so we can repeat the reproduction and debugging process against the updated code. (The upstream changes have probably invalidated the stack trace we collected above.) Thanks!
